src/typecheck.rs

```diff
@@ -1,11 +1,1384 @@
+                    if lhs_type != rhs_type {
+                        return Err(TypeError::TypeMismatch {
+                            expected: lhs_type,
+                            actual: rhs_type,
+                        });
+                    }
+                    // Allow comparison of numbers or symbols
+                    if matches!(lhs_type, Type::NumT(_) | Type::SymT) {
+                        Ok(Type::BoolT)
+                    } else {
+                        Ok(Type::BoolT)
+                    }
+                }
+                // Logical operations are always of type BoolT
+                BOp::And | BOp::Or => {
+                    if lhs_type != rhs_type {
+                        return Err(TypeError::TypeMismatch {
+                            expected: lhs_type,
+                            actual: rhs_type,
+                        });
+                    }
+                    if lhs_type == Type::BoolT {
+                        Ok(Type::BoolT)
+                    } else {
+                        Err(TypeError::TypeMismatch {
+                            expected: Type::BoolT,
+                            actual: lhs_type,
+                        })
+                    }
+                }
+            }
+        }
+        // Unary operations are always of the type of the inner expression
+        Expr::UOp { op, inner } => {
+            let inner_type = typeck_expr(inner, ctx)?;
+            match op {
+                UOp::Not => {
+                    if inner_type == Type::BoolT {
+                        Ok(Type::BoolT)
+                    } else {
+                        Err(TypeError::TypeMismatch {
+                            expected: Type::BoolT,
+                            actual: inner_type,
+                        })
+                    }
+                }
+                UOp::Negate => {
+                    if let Type::NumT(range) = inner_type {
+                        Ok(Type::NumT(range))
+                    } else {
+                        Err(TypeError::TypeMismatch {
+                            expected: Type::NumT(0..1),
+                            actual: inner_type,
+                        })
+                    }
+                }
+            }
+        }
+        // Casting is always of the type of the inner expression
+        Expr::Cast {
+            inner,
+            typ,
+            overflow: _overflow,
+        } => {
+            // First, type check the inner expression
+            let inner_type = typeck_expr(inner, ctx)?;
+
+            // According to the [T-Cast] rule, both the inner expression and the target type
+            // must be numeric types with bounds
+            match (&inner_type, typ) {
+                (Type::NumT(_), Type::NumT(_)) => {
+                    // The cast is valid, return the target type
+                    Ok(typ.clone())
+                }
+                (_, Type::NumT(_)) => {
```

Comment:
I **strongly** recommend getting rid of the whole expected type business and having a single `if matches!(...)` check. This case function should be ~5 lines of code!